### ISSUE TITLE
A plan for Basic Graph Patterns

### ISSUE THREAD

**Issue by kjetilk** [phochste/AtteanX-Store-LDF#2]
Hi!

I think the [new paper about optimizing BGPs](http://linkeddatafragments.org/publications/eswc2015.pdf) by @joachimvh , @RubenVerborgh et al could be implemented elegantly in Attean by creating a 
AtteanX::Store::LDF::Plan::BGP analogously to what I've done with the SPARQL client:
https://github.com/kjetilk/p5-atteanx-store-sparql/blob/master/lib/AtteanX/Store/SPARQL/Plan/BGP.pm
my code is a bit rough around the edges, not a whole lot of documentation and testing here yet, but the idea is that you provide a plan for the stuff that you can do better than the default implementation. Right, @kasei ?

Cheers,

Kjetil

**Comment by kasei** [comment]
Yes, that sounds right; an LDF-specific Plan class would be the way to efficiently integrate LDF BGP matching for SPARQL queries.

**Comment by RubenVerborgh** [comment]
Interesting! Any documentation or insights in how this Plan mechanism works in general, @kasei? In particular, how does the query engine know when to apply which plans?

**Comment by kasei** [comment]
Currently the query planner trusts that a plan provided by the underlying store is preferable to any equivalent plan the query planner might construct (e.g. a store-supplied BGP is better than a set of joins over multiple calls to `get_triples`). The store opts into this system by implementing the `Attean::API::CostPlanner` role.

A general overview of this planning system can be found in [Pushing Complexity Down the Stack]([http://ceur-ws.org/Vol-1268/paper14.pdf) from last year's SEMDEV workshop.

(The groundwork has been laid for not just trusting that the store-supplied plan is optimal, but instead to do a cost-based comparison of plans, but we haven't actually done the implementation work to fully support that yet.)

**Comment by RubenVerborgh** [comment]
Thanks! Will study this in more detail; I'll need something similar in JavaScript.

**Comment by kjetilk** [comment]
Creating this plan object is fairly straightforward, but I suppose that for it to be really useful, RDF::LDF should be updated to support contemporary BGP answering, as requested in https://github.com/phochste/RDF-LDF/issues/10 

So, I think I'm going to look at this issue as having a (soft) dependency on that :-)
